File: immutable.py
def qsort(xs):
    if len(xs) <= 1:
        return xs
    pivot = xs[0]
    left = [x for x in xs[1:] if x < pivot]
    right = [x for x in xs[1:] if x >= pivot]
    return qsort(left) + [pivot] + qsort(right)

if __name__ == "__main__":
    import sys
    verbose = '-v' in sys.argv
    pre_print = '-p' in sys.argv
    xs = [(i * 73 + 19) % 10000 for i in range(10000)]
    if pre_print:
        print('Before sorting:')
        print(xs)
        sys.exit(0)
    result = qsort(xs)
    if verbose:
        print(result)
    else:
        print(result[:10])

File: immutable.ml
let rec qsort = function
  | [] -> []
  | [x] -> [x]
  | pivot :: rest ->
      let left = List.filter (fun x -> x < pivot) rest in
      let right = List.filter (fun x -> x >= pivot) rest in
      qsort left @ [pivot] @ qsort right

let () =
  let verbose = Array.mem "-v" Sys.argv in
  let pre_print = Array.mem "-p" Sys.argv in
  
  (* Generate test data *)
  let xs = Array.init 10000 (fun i -> (i * 73 + 19) mod 10000) in
  let xs_list = Array.to_list xs in
  
  if pre_print then (
    print_endline "Before sorting:";
    print_string "[";
    List.iteri (fun i x ->
      if i > 0 then print_string ", ";
      print_int x
    ) xs_list;
    print_endline "]"
  ) else (
    let result = qsort xs_list in
    if verbose then (
      print_string "[";
      List.iteri (fun i x ->
        if i > 0 then print_string ", ";
        print_int x
      ) result;
      print_endline "]"
    ) else (
      let first_10 = List.filteri (fun i _ -> i < 10) result in
      print_string "[";
      List.iteri (fun i x ->
        if i > 0 then print_string ", ";
        print_int x
      ) first_10;
      print_endline "]"
    )
  )
